init.lua

```diff
@@ -1220,6 +1220,97 @@ function Form:update_where(func)
     end
 end
 
+local embed_create_ctx_mt = {}
+
+function embed_create_ctx_mt:__index(key)
+    -- rawget rensures we don't do recursion
+    --  and ensures it doesn't get wrongly prefixed
+    local form = rawget(self, "_flow_embed_parent_form")
+    local prefix = rawget(self,"_flow_embed_prefix")
+    return form[prefix .. key]
+end
+
+function embed_create_ctx_mt:__newindex(key, value)
+    -- rawget ensures it doesn't get wrongly prefixed
+    local form = rawget(self, "_flow_embed_parent_form")
+    local prefix = rawget(self,"_flow_embed_prefix")
+    form[prefix .. key] = value
+end
+
+local function embed_create_ctx(ctx, name, prefix)
+    if not ctx[name] then
+        -- Proxy ctx.form acess to the real form
+        -- We keep the metatable in memory so luaJIT can optimize indexes
+        ctx[name] = {
+            form = setmetatable(
+                {
+                    _flow_embed_prefix = prefix,
+                    _flow_embed_parent_form = ctx.form
+                },
+                embed_create_ctx_mt
+            )
+        }
+    end
+    return ctx[name]
+end
+
+local function embed_wrap_callback_func(func, name, prefix)
+    return function(player, ctx)
+        return func(player, embed_create_ctx(ctx, name, prefix))
+    end
+end
+
+local function embed_add_prefix(node, name, prefix)
+    if node.type == "style" and node.selectors then
+        -- Add prefix to style[] selectors
+        for i, selector in ipairs(node.selectors) do
+            node.selectors[i] = prefix .. selector
+        end
+    elseif node.time == "scroll_container" and node.scrollbar_name then
+        node.scrollbar_name = prefix .. node.scrollbar_name
+    elseif node.type == "tooltip" and node.gui_element_name then
+        node.gui_element_name = prefix .. node.gui_element_name
+    end
+
+    -- Add prefix to all names
+    if node.name then
+        node.name = prefix .. node.name
+    end
+
+    -- Wrap callback functions
+    if node.on_event then
+        node.on_event = embed_wrap_callback_func(node.on_event, name, prefix)
+    end
+    if node.on_quit then
+        node.on_quit = embed_wrap_callback_func(node.on_quit, name, prefix)
+    end
+
+    -- Recurse to child nodes
+    for _, child in ipairs(node) do
+        embed_add_prefix(child, name, prefix)
+    end
+end
+
+-- TODO: unit test this
+function Form:embed(fields)
+    local player = fields.player
+    local name = fields.name
+    -- TODO: it might be cool to somehow pass elements down (number-indexes of
```

Comment:
I suppose this could be a later feature request.

Reply:
Ah, sorry. May have worded it wrong.

I meant something like an API that facilitates this:

```lua
local a_form = flow.make_gui(function(player, ctx, embeded_children)
    -- the third arg here could instead be passed in via something inside ctx
    return gui.VBox(embeded_children or {})
end)
assert.same(
    gui.VBox{
        gui.Label{  label = "hello, world" }
    },
    a_form:embed{
        player = player,
        prefix = "b",
        gui.Label{ label = "hello, world" }
    }
)
```

Reply:
I'm not aware that it would be needed. A better approach is for the mod that provides `a_form` to provide an overidable function or some other API, instead of going through `Form.embed`.

I think that this deserves a case-study before it's implemented, personally. It adds API complexity and can limit future API designs.